src/database-sqlite3.cpp

```diff
@@ -62,88 +68,484 @@ SQLite format specification:
+	return mode;
+}
+
+
+void SQLite3::setJournalMode()
+{
+	sqlite3_stmt *stmt_journal;
+
+	// Determine previous mode for user feedback
+	std::string previous_mode = getDBJournalMode();
+	if (m_journal_mode == "") {
+		m_journal_mode = previous_mode;
+		infostream << "SQLite3: sqlite_journal_mode is not set in world.mt."
+			<< " Detected mode from database: " << previous_mode
+			<< std::endl;
+	}
+
+	std::string new_mode = previous_mode;
+	if (previous_mode != m_journal_mode) {
+		std::string sql_code = "PRAGMA journal_mode = ";
+		sql_code += m_journal_mode + ";";
+		SQLOK(sqlite3_prepare_v2(m_database, sql_code.c_str(), -1, &stmt_journal, NULL),
+			"SQLite3: Failed to prepare set-journal-mode query")
+		if (sqlite3_step(stmt_journal) != SQLITE_ROW)
+			throw DatabaseException("SQLite3: Failed to execute journal mode statement");
+		new_mode = lowercase((const char *)sqlite3_column_text(stmt_journal, 0));
+		SQLOK_ERRSTREAM(sqlite3_finalize(stmt_journal),
+			"SQLite3: Failed to finalize set-journal-mode statement")
+
+		if (new_mode != m_journal_mode) {
+			warningstream << "SQLite3: failed to set journal mode '"
+				<< m_journal_mode << "' on database. Mode is now: '"
+				<< new_mode << "'" << std::endl;
+		}
+		if (previous_mode != new_mode) {
+			infostream << "SQLite3: changed database journal mode from '"
+				<< previous_mode << "' to '" << new_mode << "'"
+				<< std::endl;
+		}
+		m_journal_mode = new_mode;
+	} else {
+		infostream << "SQLite3: database journal mode is '"
+			<< previous_mode << "' (not changed)" << std::endl;
+	}
+
+	if (m_journal_mode == "wal") {
+		SQLOK_ERRSTREAM(sqlite3_wal_autocheckpoint(m_database, 0),
+			"Failed to disable auto-checkpointing for WAL mode");
+	}
+}
+
+
+void SQLite3::setSynchronousLevel()
+{
+	if (m_synchronous == -1)
+		m_synchronous = (m_journal_mode == "wal") ? 1 : 2;
+	std::string query_str = std::string("PRAGMA synchronous = ")
+			 + itos(m_synchronous);
+	SQLOK(sqlite3_exec(m_database, query_str.c_str(), NULL, NULL, NULL),
+		"SQLite3: Failed to modify sqlite3 synchronous level");
+	infostream << "SQLite3: database synchronous level set to: "
+		<< m_synchronous << std::endl;
+}
+
+
+void SQLite3::startWALCheckpointThread()
+{
+	if (!m_database)
+		throw DatabaseException("SQLite3: Internal error: trying to start"
+			" WAL thread, but database is not yet open");
+
+	if (m_journal_mode != "wal" || getDBJournalMode() != "wal")
+		throw DatabaseException("SQLite3: Internal error: trying to start"
+			" WAL thread, but journal mode is not 'wal'");
+
+	if (m_walCheckpointThread)
+		throw DatabaseException("SQLite3: Internal error: trying to start"
+			" WAL thread, but it is already running");
+
+	setAutoCheckpoint(false);
+	m_walCheckpointThread = new SQLite3CheckpointThread(*this);
+	m_walCheckpointThread->start();
+}
+
+void SQLite3::stopWALCheckpointThread()
+{
+	if (m_walCheckpointThread) {
+		if (!m_walCheckpointThread->shutdown(10)) {
```

Comment:
Will change to 5. I don't want to make it too short: I want to avoid forcibly terminating the thread without it having a chance to close the database. The typical shutdown time would be a split second, and IMO the chance of the maximum ever being hit at all are probably nearly zero.